test/x509_test.c

```diff
@@ -0,0 +1,149 @@
+    0xb1, 0x86, 0xa3, 0x3b, 0x26, 0xda, 0x7e, 0x8b, 0x44, 0x45, 0xc6, 0x46, 0x14, 0x04, 0x22, 0x2b,
+    0xe5, 0x2a, 0x62, 0x84, 0xc5, 0x94, 0xa0, 0x1b, 0xaa, 0xa9
+};
+
+/* Some simple CRL data */
+static const unsigned char crldata[] = {
+    0x30, 0x81, 0x8B, 0x30, 0x31, 0x02, 0x01, 0x01, 0x30, 0x0C, 0x06, 0x08, 0x2A, 0x86, 0x48, 0xCE,
+    0x3D, 0x04, 0x03, 0x02, 0x05, 0x00, 0x30, 0x0F, 0x31, 0x0D, 0x30, 0x0B, 0x06, 0x03, 0x55, 0x04,
+    0x03, 0x0C, 0x04, 0x54, 0x65, 0x73, 0x74, 0x17, 0x0D, 0x32, 0x32, 0x31, 0x30, 0x31, 0x32, 0x30,
+    0x35, 0x33, 0x34, 0x30, 0x31, 0x5A, 0x30, 0x0C, 0x06, 0x08, 0x2A, 0x86, 0x48, 0xCE, 0x3D, 0x04,
+    0x03, 0x02, 0x05, 0x00, 0x03, 0x48, 0x00, 0x30, 0x45, 0x02, 0x20, 0x75, 0xAC, 0xA9, 0xB5, 0xFE,
+    0x63, 0x09, 0x8B, 0x57, 0x4F, 0xBB, 0xC6, 0x0C, 0xA9, 0x9A, 0x7C, 0x55, 0x89, 0xF9, 0x9C, 0x48,
+    0xE9, 0xF3, 0xED, 0xE5, 0xC2, 0x88, 0xCE, 0xEC, 0xB1, 0x51, 0xF1, 0x02, 0x21, 0x00, 0x8B, 0x93,
+    0xC5, 0xA6, 0x28, 0x48, 0x5A, 0x4E, 0x10, 0x52, 0x82, 0x12, 0x2F, 0xC4, 0x62, 0x2D, 0x3F, 0x5A,
+    0x62, 0x7F, 0x9D, 0x1B, 0x12, 0xC5, 0x36, 0x25, 0x73, 0x03, 0xF4, 0xDE, 0x62, 0x24
+};
+
+/*
+ * Test for Regression discussed in PR #19388
+ * In order for this simple test to fail, it requires the digest used for
+ * signing to be different from the alg within the loaded cert.
+ */
+static int test_x509_tsb_cache(void)
+{
+    int ret;
+    X509 *x = NULL;
+    const unsigned char *p = certdata;
+
+    ret = TEST_ptr(x = d2i_X509(NULL, &p, sizeof(certdata)))
+          && TEST_int_gt(X509_sign(x, privkey, signmd), 0)
+          && TEST_int_eq(X509_verify(x, pubkey), 1);
+    X509_free(x);
+    return ret;
+}
+
+/*
+ * Test for Regression discussed in PR #19388
+ * In order for this simple test to fail, it requires the digest used for
+ * signing to be different from the alg within the loaded cert.
+ */
+static int test_x509_crl_tsb_cache(void)
+{
+    int ret;
+    X509_CRL *crl = NULL;
+    const unsigned char *p = crldata;
+
+    ret = TEST_ptr(crl = d2i_X509_CRL(NULL, &p, sizeof(crldata)))
+          && TEST_int_gt(X509_CRL_sign(crl, privkey, signmd), 0)
+          && TEST_int_eq(X509_CRL_verify(crl, pubkey), 1);
+
+    X509_CRL_free(crl);
+    return ret;
+}
+
+int setup_tests(void)
+{
+    const unsigned char *p;
+
+    p = pubkeydata;
+    pubkey = d2i_PUBKEY(NULL, &p, sizeof(pubkeydata));
+
+    p = privkeydata;
+    privkey = d2i_PrivateKey(EVP_PKEY_EC, NULL, &p, sizeof(privkeydata));
+
+    if (pubkey == NULL || privkey == NULL) {
+        BIO_printf(bio_err, "Failed to create keys\n");
+        return 0;
+    }
+
+    /* Note this digest is different from the certificate digest */
+    signmd = EVP_MD_fetch(NULL, "SHA384", NULL);
+    if (signmd == NULL) {
+        BIO_printf(bio_err, "Failed to fetch digest\n");
+        return 0;
+    }
+
+    ADD_TEST(test_x509_tsb_cache);
+    ADD_TEST(test_x509_crl_tsb_cache);
```

Comment:
isnt this "tbs" aka. "To Be Signed" instead of tsb?

Reply:
Yes you are correct..

Reply:
updated.. thanks..